oppia/oppia.py

```diff
@@ -0,0 +1,143 @@
+# coding: utf-8
+#
+# Copyright 2015 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This XBlock embeds an instance of Oppia in the OpenEdX platform."""
+
+import pkg_resources
+
+from eventtracking import tracker
+
+from xblock.core import XBlock
+from xblock.fields import Scope, Integer, String
+from xblock.fragment import Fragment
+
+
+class OppiaXBlock(XBlock):
+    """
+    An XBlock providing an embedded Oppia exploration.
+    """
+    _EVENT_NAME_EXPLORATION_LOADED = 'oppia.exploration.loaded'
+    _EVENT_NAME_EXPLORATION_COMPLETED = 'oppia.exploration.completed'
+    _EVENT_NAME_STATE_TRANSITION = 'oppia.exploration.state.changed'
+
+    display_name = String(
+        help="Display name of the component",
+        default="Oppia Exploration",
+        scope=Scope.content)
+    oppiaid = String(
+        help="ID of the Oppia exploration to embed",
+        default="4",
+        scope=Scope.content)
+    src = String(
+        help="Source URL of the site",
+        default="https://www.oppia.org",
+        scope=Scope.content)
+
+    def resource_string(self, path):
+        """Handy helper for getting resources from our kit."""
+        data = pkg_resources.resource_string(__name__, path)
+        return data.decode("utf8")
+
+    def student_view(self, context=None):
+        """
+        The primary view of the OppiaXBlock, shown to students
+        when viewing courses.
+        """
+        html = self.resource_string("static/html/oppia.html")
+        frag = Fragment(html.format(self=self))
+        frag.add_javascript(
+            self.resource_string('static/lib/oppia-player-0.0.1-modified.js'))
+        frag.add_javascript(self.resource_string("static/js/oppia.js"))
+        frag.initialize_js('OppiaXBlock')
+        return frag
+
+    def author_view(self, context=None):
+        """
+        A view of the XBlock to show within the Studio preview. For some
+        reason, the student_view() does not display, so we show a placeholder
+        instead.
+        """
+        html = self.resource_string("static/html/oppia_preview.html")
+        frag = Fragment(html.format(self=self))
+        return frag
+
+    def _log(self, event_name, payload):
+        """
+        Logger for load, state transition and completion events.
+        """
+        tracker.emit(event_name, payload)
```

Comment:
Our analytics team recommends that you use `self.runtime.publish`, not `tracker.emit` directly.

From them: "so they are basically just bypassing the runtime and calling "tracker.emit" directly, which will work, but only in the LMS.
if you tried to run that XBlock in another xblock runtime it would fail, where as "publish" allows the runtime to define the behavior.
ideally we would expose a more comprehensive instrumentation "service" to the xblock, but we haven't gotten around to building such a thing
And more importantly, going through our runtime allows us to set some context on the tracking events being emitted."